# TRiXi: A Multi-Target Tandem Repeat-Based Method for Accurate Detection of X-Inactivation Skewing in Humans

*Dóra Goldmann<sup>1,2</sup>, Björn Gylemo<sup>1,2</sup>, Maike Bensberg<sup>1,2</sup>, Ingela Johansson<sup>1</sup>, Svenja Löffert<sup>1</sup>, Júlia Goldmann<sup>1</sup>, Lisa Haglund<sup>1</sup>, Magdalena Vitkova<sup>1</sup>, Janos Kondri<sup>1</sup>, Sandra Hellberg<sup>1</sup>, Engla Haglund<sup>1</sup>, Shadi Jafari<sup>1,3</sup>, Johnny Ludvigsson<sup>1,3</sup> & Colm E. Nestor<sup>1,3</sup>*

<sup>1</sup> Crown Princess Victoria Children´s Hospital, Department of Biomedical and Clinical Sciences (BKV), Linköping University, Linköping, Sweden

<sup>2</sup> joint first authorship
<sup>3</sup> joint senior authorship

Corresponding author: colm.nestor@liu.se

##

Here you can find scripts used to perform analyses and generate figures in the above publication (link to be added).

## Abstract
Preferential usage of either X-chromosome, skewed X-inactivation (sXCI), can profoundly alter the penetrance and expressivity of X-linked traits in females, often with life-threatening consequences. 
Despite its clinical relevance, the true frequency, origins, and stability of sXCI in human tissues remain poorly understood, in part due to a lack of effective and sensitive methods for assessing XCI status. 
Here, we present TRiXi (Tandem Repeat-based Identification of X-Inactivation), a simple and highly sensitive technique capable of detecting sXCI from as little as 10 ng of archived human DNA. 
Applying TRiXi to over 1,000 neonatal cord blood samples from healthy infants, we demonstrate that extreme sXCI is a common constitutional feature in humans. 
Use of longitudinal samples from the same donors revealed that sXCI patterns in skewed neonates were stably maintained into adolescence, whereas in non-skewed females, patterns fluctuated considerably over time. 
Using low-pass short-read sequencing, whole-genome long-read sequencing, and targeted exon sequencing, we exclude most known genetic drivers of sXCI in extremely skewed infants, supporting stochastic processes as the primary origin of skewed XCI. 
Further, by inferring X-inactivation patterns in 4,571 adult tissues from 264 females in the GTEx cohort, we demonstrate that skewed XCI is not restricted to blood, revealing widespread multi-organ sXCI in females. 
Our study represents the most comprehensive analysis of constitutional sXCI to date and reveals sXCI as a widespread and stable epigenetic trait. 
This can profoundly modify the penetrance of X-linked conditions with immediate implications for genetic diagnostics and our understanding of human disease variability. 
TRiXi provides a convenient and accessible approach to inform diagnosis of X-linked traits by assaying sXCI. Studying individuals with sXCI may yield critical insights into (1) 
the initiation of XCI and its underlying genomic elements, (2) previously unrecognized lethal X-linked traits and (3) improved polygenic risk prediction through incorporation of X-chromosome variation.
##


### Session info
*R version 4.5.0 (2025-04-11)*

| Package | Version | Source |
| :------ | :------ | :----- |
| `abind` | 1.4-8 | CRAN  |
| `ade4` | 1.7-23 | CRAN  |
| `AnnotationDbi` | 1.70.0 | Bioconductor  |
| `AnnotationFilter` | 1.32.0 | Bioconductor  |
| `assertthat` | 0.2.1 | CRAN  |
| `backports` | 1.5.0 | CRAN  |
| `bamsignals` | 1.40.0 | Bioconductor  |
| `base64enc` | 0.1-3 | CRAN  |
| `beachmat` | 2.24.0 | Bioconductor  |
| `beeswarm` | 0.4.0 | CRAN  |
| `bezier` | 1.1.2 | CRAN  |
| `Biobase` | 2.68.0 | Bioconductor  |
| `BiocFileCache` | 2.16.2 | Bioconductor  |
| `BiocGenerics` | 0.54.0 | Bioconductor  |
| `BiocIO` | 1.18.0 | Bioconductor  |
| `BiocParallel` | 1.42.0 | Bioconductor  |
| `BiocSingular` | 1.24.0 | Bioconductor  |
| `biomaRt` | 2.64.0 | Bioconductor  |
| `Biostrings` | 2.76.0 | Bioconductor  |
| `biovizBase` | 1.56.0 | Bioconductor  |
| `bit` | 4.6.0 | CRAN  |
| `bit64` | 4.6.0-1 | CRAN  |
| `bitops` | 1.0-9 | CRAN  |
| `blob` | 1.2.4 | CRAN  |
| `broom` | 1.0.8 | CRAN  |
| `BSgenome` | 1.76.0 | Bioconductor  |
| `BSgenome.Hsapiens.UCSC.hg38` | 1.4.5 | Bioconductor  |
| `bsseq` | 1.44.1 | Bioconductor  |
| `cachem` | 1.1.0 | CRAN  |
| `car` | 3.1-3 | CRAN  |
| `carData` | 3.0-5 | CRAN  |
| `CGHbase` | 1.68.0 | Bioconductor  |
| `CGHcall` | 2.70.0 | Bioconductor  |
| `checkmate` | 2.3.2 | CRAN  |
| `class` | 7.3-23 | CRAN  |
| `cli` | 3.6.5 | CRAN  |
| `cluster` | 2.1.8.1 | CRAN  |
| `codetools` | 0.2-19 | CRAN  |
| `colorspace` | 2.1-1 | CRAN  |
| `cowplot` | 1.2.0 | CRAN  |
| `cpp11` | 0.5.2 | CRAN  |
| `crayon` | 1.5.3 | CRAN  |
| `curl` | 6.2.3 | CRAN  |
| `data.table` | 1.17.4 | CRAN  |
| `DBI` | 1.2.3 | CRAN  |
| `dbplyr` | 2.5.0 | CRAN  |
| `dbscan` | 1.2.3 | CRAN  |
| `DelayedArray` | 0.34.1 | Bioconductor  |
| `DelayedMatrixStats` | 1.30.0 | Bioconductor  |
| `dichromat` | 2.0-0.1 | CRAN  |
| `digest` | 0.6.37 | CRAN  |
| `DNAcopy` | 1.82.0 | Bioconductor  |
| `doParallel` | 1.0.17 | CRAN  |
| `dplyr` | 1.1.4 | CRAN  |
| `e1071` | 1.7-16 | CRAN  |
| `ensembldb` | 2.32.0 | Bioconductor  |
| `evaluate` | 1.0.3 | CRAN  |
| `farver` | 2.1.2 | CRAN  |
| `fastmap` | 1.2.0 | CRAN  |
| `filelock` | 1.0.3 | CRAN  |
| `forcats` | 1.0.0 | CRAN  |
| `foreach` | 1.5.2 | CRAN  |
| `foreign` | 0.8-90 | CRAN  |
| `Formula` | 1.2-5 | CRAN  |
| `fs` | 1.6.6 | CRAN  |
| `future` | 1.49.0 | CRAN  |
| `future.apply` | 1.11.3 | CRAN  |
| `generics` | 0.1.4 | CRAN  |
| `GenomeInfoDb` | 1.44.0 | Bioconductor  |
| `GenomeInfoDbData` | 1.2.14 | Bioconductor  |
| `GenomicAlignments` | 1.44.0 | Bioconductor  |
| `GenomicFeatures` | 1.60.0 | Bioconductor  |
| `GenomicRanges` | 1.60.0 | Bioconductor  |
| `ggalluvial` | 0.12.5 | CRAN  |
| `ggbeeswarm` | 0.7.2 | CRAN  |
| `ggplot2` | 3.5.2 | CRAN  |
| `ggpubr` | 0.6.1 | CRAN  |
| `ggrastr` | 1.0.2 | CRAN  |
| `ggrepel` | 0.9.6 | CRAN  |
| `ggridges` | 0.5.7 | CRAN  |
| `ggsignif` | 0.6.4 | CRAN  |
| `globals` | 0.18.0 | CRAN  |
| `glue` | 1.8.0 | CRAN  |
| `gridExtra` | 2.3 | CRAN  |
| `gtable` | 0.3.6 | CRAN  |
| `gtools` | 3.9.5 | CRAN  |
| `h5mread` | 1.0.1 | Bioconductor  |
| `HDF5Array` | 1.36.0 | Bioconductor  |
| `Hmisc` | 5.2-3 | CRAN  |
| `hms` | 1.1.3 | CRAN  |
| `htmlTable` | 2.4.3 | CRAN  |
| `htmltools` | 0.5.8.1 | CRAN  |
| `htmlwidgets` | 1.6.4 | CRAN  |
| `httr` | 1.4.7 | CRAN  |
| `httr2` | 1.1.2 | CRAN  |
| `impute` | 1.82.0 | Bioconductor  |
| `IRanges` | 2.42.0 | Bioconductor  |
| `irlba` | 2.3.5.1 | CRAN  |
| `ismev` | 1.43 | CRAN  |
| `iterators` | 1.0.14 | CRAN  |
| `jsonlite` | 2.0.0 | CRAN  |
| `karyoploteR` | 1.34.2 | Bioconductor  |
| `KEGGREST` | 1.48.1 | Bioconductor  |
| `knitr` | 1.5 | CRAN  |
| `lattice` | 0.22-5 | CRAN  |
| `lazyeval` | 0.2.2 | CRAN  |
| `lifecycle` | 1.0.4 | CRAN  |
| `limma` | 3.64.1 | Bioconductor  |
| `listenv` | 0.9.1 | CRAN  |
| `locfit` | 1.5-9.12 | CRAN  |
| `lubridate` | 1.9.4 | CRAN  |
| `magrittr` | 2.0.3 | CRAN  |
| `marray` | 1.86.0 | Bioconductor  |
| `MASS` | 7.3-65 | CRAN  |
| `Matrix` | 1.7-3 | CRAN  |
| `MatrixGenerics` | 1.20.0 | Bioconductor  |
| `matrixStats` | 1.5.0 | CRAN  |
| `memoise` | 2.0.1 | CRAN  |
| `mgcv` | 1.9-1 | CRAN  |
| `NanoMethViz` | 3.4.0 | Bioconductor  |
| `nlme` | 3.1-168 | CRAN  |
| `nnet` | 7.3-20 | CRAN  |
| `parallelly` | 1.44.0 | CRAN  |
| `patchwork` | 1.3.2 | CRAN  |
| `permute` | 0.9-8 | CRAN  |
| `pillar` | 1.10.2 | CRAN  |
| `pkgconfig` | 2.0.3 | CRAN  |
| `plyr` | 1.8.9 | CRAN  |
| `png` | 0.1-8 | CRAN  |
| `prettyunits` | 1.2.0 | CRAN  |
| `progress` | 1.2.3 | CRAN  |
| `ProtGenerics` | 1.40.0 | Bioconductor  |
| `proxy` | 0.4-27 | CRAN  |
| `purrr` | 1.0.4 | CRAN  |
| `QDNAseq` | 1.44.0 | Bioconductor  |
| `QDNAseq.hg38` | 1.1.0 | Github  |
| `R.methodsS3` | 1.8.2 | CRAN  |
| `R.oo` | 1.27.1 | CRAN  |
| `R.utils` | 2.13.0 | CRAN  |
| `R6` | 2.6.1 | CRAN  |
| `rappdirs` | 0.3.3 | CRAN  |
| `RColorBrewer` | 1.1-3 | CRAN  |
| `Rcpp` | 1.0.14 | CRAN  |
| `RCurl` | 1.98-1.17 | CRAN  |
| `readr` | 2.1.5 | CRAN  |
| `regioneR` | 1.40.1 | Bioconductor  |
| `restfulr` | 0.0.15 | CRAN  |
| `rhdf5` | 2.52.1 | Bioconductor  |
| `rhdf5filters` | 1.20.0 | Bioconductor  |
| `Rhdf5lib` | 1.30.0 | Bioconductor  |
| `rjson` | 0.2.23 | CRAN  |
| `rlang` | 1.1.6 | CRAN  |
| `rmarkdown` | 2.29 | CRAN  |
| `rpart` | 4.1.24 | CRAN  |
| `Rsamtools` | 2.24.0 | Bioconductor  |
| `RSQLite` | 2.4.1 | CRAN  |
| `rstatix` | 0.7.2 | CRAN  |
| `rstudioapi` | 0.17.1 | CRAN  |
| `rsvd` | 1.0.5 | CRAN  |
| `rtracklayer` | 1.68.0 | Bioconductor  |
| `S4Arrays` | 1.8.0 | Bioconductor  |
| `S4Vectors` | 0.46.0 | Bioconductor  |
| `ScaledMatrix` | 1.16.0 | Bioconductor  |
| `scales` | 1.4.0 | CRAN  |
| `scanstatistics` | 1.1.1 | CRAN  |
| `seqinr` | 4.2-36 | CRAN  |
| `sessioninfo` | 1.2.3 | CRAN  |
| `sets` | 1.0-25 | CRAN  |
| `SparseArray` | 1.8.0 | Bioconductor  |
| `sparseMatrixStats` | 1.20.0 | Bioconductor  |
| `statmod` | 1.5.0 | CRAN  |
| `stringi` | 1.8.7 | CRAN  |
| `stringr` | 1.5.1 | CRAN  |
| `SummarizedExperiment` | 1.38.1 | Bioconductor  |
| `tibble` | 3.2.1 | CRAN  |
| `tidyr` | 1.3.1 | CRAN  |
| `tidyselect` | 1.2.1 | CRAN  |
| `tidyverse` | 2.0.0 | CRAN  |
| `timechange` | 0.3.0 | CRAN  |
| `tzdb` | 0.5.0 | CRAN  |
| `UCSC.utils` | 1.4.0 | Bioconductor  |
| `VariantAnnotation` | 1.54.1 | Bioconductor  |
| `vctrs` | 0.6.5 | CRAN  |
| `vipor` | 0.4.7 | CRAN  |
| `withr` | 3.0.2 | CRAN  |
| `xfun` | 0.52 | CRAN  |
| `XML` | 3.99-0.18 | CRAN  |
| `xml2` | 1.3.8 | CRAN  |
| `XVector` | 0.48.0 | Bioconductor  |
| `yaml` | 2.3.10 | CRAN  |
| `zlibbioc` | 1.54.0 | Bioconductor  |
